Spigot-API-Patches/0082-Basic-PlayerProfile-interface.patch

```diff
@@ -0,0 +1,224 @@
++     */
++    @Nullable UUID getId();
++
++    /**
++     * @return A copy of this players properties, such as textures.
++     * Values specified here are subject to implementation details.
++     */
++    @Nonnull Set<ProfileProperty> getProperties();
++
++    /**
++     * Sets a property. If the property already exists, the previous one will be replaced
++     * @param property Property to set.
++     */
++    void setProperty(ProfileProperty property);
++
++    /**
++     * Sets multiple properties. If any of the set properties already exist, it will be replaced
++     * @param properties The properties to set
++     */
++    void setProperties(Collection<ProfileProperty> properties);
++
++    /**
++     * Removes a specific property from this profile
++     * @param property The property to remove
++     * @return If a property was removed
++     */
++    boolean removeProperty(String property);
++
++    /**
++     * Removes a specific property from this profile
++     * @param property The property to remove
++     * @return If a property was removed
++     */
++    default boolean removeProperty(@Nonnull ProfileProperty property) {
++        return removeProperty(property.getName());
++    }
++
++    /**
++     * Clears all properties on this profile
++     */
++    void clearProperties();
++
++    /**
++     * @return Does this profile have Name, UUID and Textures filled in
++     */
++    boolean isComplete();
++}
+diff --git a/src/main/java/com/destroystokyo/paper/profile/ProfileProperty.java b/src/main/java/com/destroystokyo/paper/profile/ProfileProperty.java
+new file mode 100644
+index 00000000..367a75d6
+--- /dev/null
++++ b/src/main/java/com/destroystokyo/paper/profile/ProfileProperty.java
+@@ -0,0 +1,47 @@
++package com.destroystokyo.paper.profile;
++
++import java.util.Objects;
++
++public class ProfileProperty {
++    private final String name;
++    private final String value;
++    private final String signature;
++
++    public ProfileProperty(String name, String value) {
++        this.name = name;
++        this.value = value;
++        this.signature = null;
++    }
++    public ProfileProperty(String name, String value, String signature) {
++        this.name = name;
++        this.value = value;
++        this.signature = signature;
++    }
++
++    public String getName() {
++        return name;
++    }
++
++    public String getValue() {
++        return value;
++    }
++
++    public String getSignature() {
++        return signature;
++    }
++
++    @Override
++    public boolean equals(Object o) {
++        if (this == o) return true;
++        if (o == null || getClass() != o.getClass()) return false;
++        ProfileProperty that = (ProfileProperty) o;
++        return Objects.equals(name, that.name) &&
++                Objects.equals(value, that.value) &&
++                Objects.equals(signature, that.signature);
++    }
++
++    @Override
++    public int hashCode() {
++        return Objects.hash(name);
```

Comment:
What if the name isn't set? I think this should check if the whether the ID or the name is set.

Reply:
A ProfileProperty requires a name, I guess I need to add the NonNull annotation to that and Preconditions